### DISCUSSION TITLE
How to Reclaim Disk Space for Deleted and Unused Images

### DISCUSSION THREAD

**Discussion by noelantoniooo** [rancher-sandbox/rancher-desktop#2871] [Q&A]
Hello,

I tried to delete unused images via rancher desktop UI. But I've noticed that even though I've already deleted it, I didn't reclaim disk space. But when I reset my entire rancher desktop configuration, it freed up all space back.

**Comment by adamkpickering** [comment]
I would need more info to know what to do for your case. Are the images being used by `docker` or `nerdctl`? If so, did you try doing `docker image prune` or `nerdctl image prune`? But ultimately, the most reliable way to clean things up is a factory reset.

**Comment by cmahnke** [comment]
This issue still persists at least with version 1.8.1 on Mac OS 13.3.
I suspect some internal disk image which is only resized at a restart. Maybe this is something to create an issue on?
I haven't looked yet, if there already is one.

**Reply by jandubois** [reply]
This is correct; disk space from the VM is only returned to the host filesystem when Rancher Desktop is restarted. I guess we could run `fstrim` after deleting images via the GUI, but there isn't really a mechanism when images are removed from the CLI.

**Reply by cmahnke** [reply]
How about writing a feature request to change this?
I can see several workarounds:
* Some sort of Cron Job
* Some helper deamon or a privileged pod
* A button as part of the GUI (maybe hidden as "advanced / experimental feature")

Anyhow, a nice change would be to run the fstrim on shutdown of the application, not on startup since it can really be a problem when the disk is full: One can't start Rancher Desktop in this case, since it fails on start with a full disk.

**Reply by jandubois** [reply]
> Anyhow, a nice change would be to run the fstrim on shutdown of the application, not on startup since it can really be a problem when the disk is full: One can't start Rancher Desktop in this case, since it fails on start with a full disk.

This is already the case. `fstrim` is run during shutdown: https://github.com/rancher-sandbox/rancher-desktop/blob/542da9d01a49dc33e89cb9d758f4f7870398eff2/pkg/rancher-desktop/backend/lima.ts#L2049

And again during startup (just in case Rancher Desktop threw an error during shutdown and did not execute the `fstrim`).

I don't see the point of running it continuously; if you want to return the space to the host and not just reuse it inside the VM, why not shutdown the app to do it?

If you really want to, you can trigger it from the commandline:

```
rdctl shell sudo fstrim /mnt/data
```

**Reply by cmahnke** [reply]
I'm sorry, im not familiar with the source, can you confirm that it's alway run on shutdown, even if the disk is full?
From my experience I would doubt that. It certainly throws some error, as you suspected.

Whether it makes sense or not depends on the use-case: I user Rancher Desktop mainly to build and test Docker images, as a drop in replacement for Docker Desktop and to test K8s deployments. This produces many half baked builds and pollutes the Builder cache. And thus is almost always growing.
Actually I don't see the point in restarting the application since this is a) quite slow (about 120 secs on a M1 Max, 32GB) and b) there is a restart option missing in the context menu / GUI 😉

But many thanks for the provided command, it does the trick!

**Reply by cmahnke** [reply]
There seems to be a regression: The above stoped to work. I accidently commented on the wrong issue #2398 it's more certainly related to #2285.
I noticed it with 1.15.0 but given the current size of `lima/0/diffdisk`the issue has to been reintroduced earlier and slipped thru QA.

Annoyingly the reset button also seem to have vanished - maybe a candidate for a hotfix?

**Reply by jandubois** [reply]
> There seems to be a regression:

I'm not aware of any regression; Rancher Desktop should continue to call `fstrim` on start and exit.

Note that you have to use `du diffdisk` to see how much space it takes, as it is a sparse file with unallocated blocks. The `ls diffdisk` size show the file size, but not the disk usage.

> Annoyingly the reset button also seem to have vanished

Are you talking about the "Factory Reset" button on the Troubleshooting page? It is still there. Can you send a screenshot? Note that you can also run `rdctl factory-reset`, but there should be no changes in that part of the UI.

**Reply by cmahnke** [reply]
Ok, great, you're right about the button, I was expecting it to be in the settings and could have sworn that once have been there.

Anyways, I wasn't exclusively to startup and stopping, but `rdctl shell sudo fstrim /mnt/data` - this doesn't seem to work.
In the mean time I just deleted the `diffdisk` file, it freed much space, I lost the images (that was expected) and it was transparently recreated...

**Reply by sluglit** [reply]
For Windows 
See #https://github.com/rancher-sandbox/rancher-desktop/issues/2398#issuecomment-2516235321
